src/core/tracking.js

```diff
@@ -83,7 +83,9 @@ export class Tracking {
   }
 
   log(...args) {
-    this._log.info(this.logPrefix, ...args);
+    if (this._log) {
```

Comment:
This change, I guess, needs a test case.

Reply:
I agree.  I tried to figure out how to write a test for this, and it's pretty confusing. I think we'd have to use DI for `this._log`, but then that would generally default `this._log` to `log` and therefore we'd no longer be able to check for `this._log`'s existence in the `log` function. I suppose we could reset `this._log` to undefined if `window` is undefined in the constructor, but this all starts to sound like a lot of unnecessary work.

Maybe this approach isn't the best fix. Perhaps it would be better to not `return` at line 47 in the constructor when `typeof window === 'undefined'`, and allow the code to continue to run, and then set `this.enabled = false` if `typeof window === 'undefined'`. I think that would make testing simpler. What do you think @willdurand?

Reply:
if you `null` the parameter, isn't it supposed to work? the default value should not be used because it is not undefined, but the condition should be fired, no?

Reply:
Maybe I didn't describe the problem very well. The code we're trying to test is:

```js
  log(...args) {
    if (this._log) {
      this._log.info(this.logPrefix, ...args);
    }
  }
```

and the code path we want to test is where `this._log` is falsey. *But* in order to verify that `this._log.info()` was not called, we need `this._log` to be a mock, which is not falsey. So I don't see how it's possible to test this exact code, which suggests to me that it's not a good design.

This is a small enough issue that I'd be happy to just let it go, knowing it works because tests that failed without it pass with it, but looking at the code for `tracking.js` it also seems like not the best approach. I thought maybe the approach I suggested above (which involves not bailing out of the constructor immediately if `window` is undefined) might be better. It's not really a requirement that we not log when `window` doesn't exist. I just added that "fix" because it addressed the issue we had at the time, but that issue was based on the current implementation of `Tracking` which I now feel is sub-optimal.

What do you think?

We could also land this as is and file a follow-up bug to address the missing coverage by improving upon `Tracking`.

Reply:
Ah no, sorry it is me. I skipped the paragraph about the constructor being weird..

I am not a fan of landing things and fix later because we might never come back.

Looking at the code and the tests, if we inject `_log` in the constructor (`_log = log`), can't we smoke test `log()` and ensures it does not output an error? I dont think we can assert on `this._log.info()` but we could maybe invert the expectation and assert for no error when there is no `this._log`. That would cover your change.

Reply:
Ok I added a smoke test for this. It's a bit weird but I guess it does the trick.